diff --git a/doc/Makefile.am b/doc/Makefile.am
index 9346696..f021999 100644
--- a/doc/Makefile.am
+++ b/doc/Makefile.am
@@ -10,7 +10,7 @@ AM_CPPFLAGS = -I. -I.. -I=$(includedir) -I$(top_srcdir)
 
 MAKEDOC_CPP = $(CPP) $(AM_CPPFLAGS) $(DEFS) $(CPPFLAGS) -D_MAKEDOC -C
 
-noinst_PROGRAMS = makedoc
+makedoc : $(makedoc_SOURCES)
 
 EXTRA_DIST = dotlock.man		\
         smime_keys.man                  \
diff --git a/doc/makedoc.c b/doc/makedoc.c
index 1fcbc24..c5ca6e3 100644
--- a/doc/makedoc.c
+++ b/doc/makedoc.c
@@ -51,7 +51,7 @@
 #ifndef HAVE_STRERROR
 #ifndef STDC_HEADERS
 extern int sys_nerr;
-extern char *sys_errlist[];
+/*extern char *sys_errlist[];*/
 #endif
 
 #define strerror(x) ((x) > 0 && (x) < sys_nerr) ? sys_errlist[(x)] : 0
