From 209321a4ff51cf2e1ffa493a4dd01c0a01899b08 Mon Sep 17 00:00:00 2001
From: Mark Asselstine <mark.asselstine@windriver.com>
Date: Tue, 14 Aug 2018 11:16:09 -0400
Subject: [PATCH] Create block devs containers

Signed-off-by: Jason Wessel <jason.wessel@windriver.com>
Signed-off-by: Mark Asselstine <mark.asselstine@windriver.com>
---
 rules/50-udev-default.rules.in | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/rules/50-udev-default.rules.in b/rules/50-udev-default.rules.in
index 191f56f..2c5cf7a 100644
--- a/rules/50-udev-default.rules.in
+++ b/rules/50-udev-default.rules.in
@@ -56,7 +56,8 @@ KERNEL=="lp[0-9]*", GROUP="lp"
 KERNEL=="irlpt[0-9]*", GROUP="lp"
 SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ENV{ID_USB_INTERFACES}=="*:0701??:*", GROUP="lp"
 
-SUBSYSTEM=="block", GROUP="disk"
+# if /dev is not devtmpfs, work around it by creating the device that is required for hotplug
+SUBSYSTEM=="block", GROUP="disk", IMPORT{program}="/bin/sh -c 'if [ ! -e %N ] ; then /bin/mknod %N b %M %m; fi'"
 SUBSYSTEM=="block", KERNEL=="sr[0-9]*", GROUP="cdrom"
 SUBSYSTEM=="scsi_generic", SUBSYSTEMS=="scsi", ATTRS{type}=="4|5", GROUP="cdrom"
 KERNEL=="sch[0-9]*", GROUP="cdrom"
-- 
2.7.4

