diff --git a/src/boot/grub2/ostree-grub-generator b/src/boot/grub2/ostree-grub-generator
index 82e66bd..b9cbe22 100644
--- a/src/boot/grub2/ostree-grub-generator
+++ b/src/boot/grub2/ostree-grub-generator
@@ -24,7 +24,16 @@ set -e
 script=$(basename ${0})
 # Atomically safe location where to generete grub.cfg when executing system upgrade.
 new_grub2_cfg=${2}
-entries_path=$(dirname $new_grub2_cfg)/entries
+#entries_path=$(dirname $new_grub2_cfg)/entries
+if [ -n "$_OSTREE_GRUB2_BOOTVERSION" -a -d /boot/loader.${_OSTREE_GRUB2_BOOTVERSION}/entries ]; then
+    entries_path="/boot/loader.${_OSTREE_GRUB2_BOOTVERSION}/entries"
+else
+    if [ -d $(dirname $new_grub2_cfg)/../../../loader.${_OSTREE_GRUB2_BOOTVERSION}/entries ]; then
+        entries_path=$(dirname $new_grub2_cfg)/../../../loader.${_OSTREE_GRUB2_BOOTVERSION}/entries
+    else
+        entries_path=$(dirname $new_grub2_cfg)/entries
+    fi
+fi
 
 read_config()
 {
@@ -92,6 +101,7 @@ cat >> ${new_grub2_cfg} <<EOF
 serial --unit=0 --speed=115200 --word=8 --parity=no --stop=1
 default=boot
 timeout=10
+set root=(hd0,msdos2)
 
 EOF
 }
