From 792b602066aa53bc1970fe86bae7c9879ffbc24a Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Sat, 11 Apr 2015 00:11:03 -0400
Subject: [PATCH] polkit: use  instead of hardcoded /lib

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 data/Makefile.am              | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: polkit-0.112/data/Makefile.am
===================================================================
--- polkit-0.112.orig/data/Makefile.am
+++ polkit-0.112/data/Makefile.am
@@ -2,7 +2,7 @@
 
 NULL =
 
-libprivdir = $(prefix)/lib/polkit-1
+libprivdir = $(libdir)/polkit-1
 
 # ----------------------------------------------------------------------------------------------------
 
